packages/orm/src/dataloaders/findOrCreateDataLoader.ts

```diff
@@ -17,27 +17,44 @@ export function findOrCreateDataLoader<T extends Entity>(
   type: EntityConstructor<T>,
   where: Partial<OptsOf<T>>,
   softDeletes: "include" | "exclude",
-): DataLoader<Key<T>, T> {
+): DataLoader<FindOrCreateKey<T>, T> {
   const meta = getMetadata(type);
 
   // Do some extra logic to make citext not create duplicate entities
   const hasAnyCitext = Object.values(meta.allFields).some((f) => f.kind === "primitive" && f.citext);
-  const cacheKeyFn: (key: Key<T>) => Key<T> = !hasAnyCitext
-    ? whereFilterHash
-    : (key) =>
-        whereFilterHash({
-          where: maybeLower(meta, key.where),
-          ifNew: maybeLower(meta, key.ifNew as any),
-          upsert: maybeLower(meta, key.upsert),
-        } as any) as Key<T>;
+
+  const cacheKeyFn: (key: FindOrCreateKey<T>) => FindOrCreateKey<T> =
+    // If no citext, we can use our regular `whereFilterHash` that replaces `Author(...)` with `id=1`
+    !hasAnyCitext
+      ? whereFilterHash
+      : // Otherwise rewrite each of the `{ firstName: "AaA" }` values in each of where/ifNew/upsert hashes
+        (key) =>
+          whereFilterHash({
+            where: maybeLower(meta, key.where),
+            ifNew: maybeLower(meta, key.ifNew as any),
+            upsert: maybeLower(meta, key.upsert),
+          } as any) as FindOrCreateKey<T>;
 
   // Use `whereFilterHash` to batch the same `findOrCreate` `where: { firstName: "a1" }` calls together
   // to avoid creating duplicates. Also use `whereFilterHash` b/c if a new entity is included in `where`,
   // it will use `entity.toString()` to keep it unique from other new entities.
   const whereValue = whereFilterHash(!hasAnyCitext ? (where as any) : (maybeLower(meta, where) as any));
   const batchKey = `${type.name}-${whereValue}-${softDeletes}`;
 
-  return em.getLoader<Key<T>, T>(
+  const invalidKeys = Object.keys(where).filter((key) => {
+    const field = meta.allFields[key];
+    const supported =
+      field.kind === "primitive" || field.kind === "enum" || field.kind === "m2o" || field.kind === "poly";
```

Comment:
thought: Could be nice to make an array of these valid keys and use `validKeys.includes(field.kind)` here.

Then the error message could also use the validKeys to build the message.

Granted, it is not likely the validKeys will be changing much in the future 🤷